### DISCUSSION TITLE
✅ Access-Control-Allow-Headers allow "*"

### DISCUSSION THREAD

**Discussion by HardCoreQual** [RobinTail/express-zod-api#500] [Ideas]
in experss-zod-api has
`
    response.set("Access-Control-Allow-Headers", "content-type");
`
but here also can be send and authorization  for example, I think better solution is to have possibility send all cors settings in Config

**Comment by RobinTail** [comment]
I presume you're talking about this line:

https://github.com/RobinTail/express-zod-api/blob/master/src/endpoint.ts#L138

I agree. There could be more headers.

> to have possibility send all cors settings in Config

What could these CORS setting look like in your opinion?

Should they configure these headers:
https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS#the_http_response_headers
? Or should there be something more than that?

@HardCoreQual 

P.S. Thank you for your finding

**Comment by RobinTail** [comment]
@HardCoreQual , please check out my PR #503 .
Would this be enough to fulfill your needs?

**Reply by HardCoreQual** [reply]
Yes, thanks, I temporary switch project from `createServer` to `attachRouting`, but after this update, I can revert it :)

Can be bug that
`ExpressZodEndpoidFactory.addExpressMiddleware(cors()) `  don't  work         // createServer
and is necessary to use `expressApp.use(cors())`                                               // attachRouting
or it work like expected ?

**Reply by RobinTail** [reply]
> Yes, thanks, I temporary switch project from createServer to attachRouting, but after this update, I can revert it :)

great, then I'll merge and release it soon.
Thank you.

> Can be bug that
ExpressZodEndpoidFactory.addExpressMiddleware(cors()) don't work // createServer
and is necessary to use expressApp.use(cors()) // attachRouting
or it work like expected ?

the factory's method `addExpressMiddleware()` (and it's alias `use()`) add middleware to the endpoint.
Currently all these middlewares are executed before the endpoint's handler.
However, the `OPTIONS` request (which is the part of the CORS workflow) is not being handled by the endpoint's handler.
Its case is special in the current implementation and completely controlled by the `cors` option in the config.
So, yes, I think it's something that should be improved as well by running middlewares for the `OPTIONS` request.
But I need to figure out the best way to do it.
Thank you for reporting this. I'll make a separate issue out of it. @HardCoreQual

**Reply by RobinTail** [reply]
Fixed in 7.6.1 🚀 @HardCoreQual

**Comment by RobinTail** [comment]
The feature released in version 7.5.0 🚀

**Comment by pycanis** [comment]
Hi, I'm using `express-zod-api` as middleware to express. I'm handling cors at global level with [cors](https://www.npmjs.com/package/cors) middleware before `attachRouting`. For one specific endpoint, I need to relax the restrictions. I've tried this:

```
export const corsRelaxedEndpoint = jsonEndpointsFactory
  .addExpressMiddleware(cors({ origin: '*' }))
  .build({
    method: 'post',
...
```

but it doesn't seem to work. Any idea, please? @RobinTail

**Reply by RobinTail** [reply]
does not work how, @pycanis ?
what do you see in place of that star in the response headers instead?

**Reply by pycanis** [reply]
I don't see anything, it's a third party doing the request to my endpoint.

If I use `origin: '*'` with the global middleware, it's working fine. If specify allowed origins with the global middleware and use `.addExpressMiddleware(cors({ origin: '*' }))`, it's not working.

In the logs I see my server returned `204 No Content` to `OPTIONS` request, but `POST` never comes.

I've tried to simulate the same request using [Insomnia](https://insomnia.rest/). With `POST`, I can see `Access-Control-Allow-Origin | *` but it's not there for `OPTIONS` request. So it seems like the `addExpressMiddleware` is not applied for `OPTIONS` request.

**Reply by RobinTail** [reply]
> So it seems like the addExpressMiddleware is not applied for OPTIONS request.

It's not applied when `cors: false` in your configuration.
But when it's `true`, the framework handles CORS on its own, informing on the supported methods it is aware about.
There is also a way to set `cors` with a function that would return the headers or alter the default ones.
https://ez.robintail.cz/v25.5.2/cross-origin-resource-sharing

Since you're having the `cors()` middleware globally and most likely having `cors: false` in config, I suggest you to wrap your global `cors()` into a middleware-like function that would preconfigure `cors()` depending on request in a relaxed or a strict way, @pycanis 


```ts
app.use((req, res, next) => cors({ origin: req.path.match(/certain/) ? "*" : list, ...otherOptions })(req, res, next) );
```

**Reply by pycanis** [reply]
That works, thank you!
